migrations/migrations/0019_calc_mediafile_is_public.py

```diff
@@ -0,0 +1,202 @@
+from collections import defaultdict
+from typing import Any, Dict, List, Optional, Set, Tuple, cast
+
+from datastore.migrations import (
+    BaseEvent,
+    BaseMigration,
+    CreateEvent,
+    ListUpdateEvent,
+    UpdateEvent,
+)
+from datastore.shared.util import collection_and_id_from_fqid
+
+
+class Migration(BaseMigration):
+    """
+    This migration should check for complete mediafile fields "is_public" and
+    "inherited_access_group_ids" and should fill them.
+    Strategy:
+        1. Build a dict with fqid and data in migrate_event
+        2. get_additional_events: Run thru the dict and
+           follow the parent_id-chain. Store the highest id of dict
+           in each chain in root-list, not it's parent
+        3. Go thru the root-list traversing the tree recursively, calculate
+           is public and inherited_access_group_ids and generate update requests
+           for the 2 fields, when the values differ. Stop the down recursion,
+           if the current one isn't changed and the child is not dict of changed mediafiles
+        4. Changing group_ids in the inherited_access_group_ids of a mediafile also have
+           to be reflected in the group-events
+
+    The algorithm assumes, that the events for a SINGLE mediafile entry are sent in
+    correct sequence, because this is guaranteed by the way, the write_requests of
+    actions are built.
+    Furthermore only the Create-, Update- and ListUpdateEvent are tracked,
+    the others do not influence the result
+    """
+
+    target_migration_index = 20
+
+    def position_init(self) -> None:
+        self.mediafiles: Dict[int, Dict[str, Any]] = defaultdict(dict)
+        self.events: List[BaseEvent] = []
+
+    def migrate_event(self, event: BaseEvent) -> Optional[List[BaseEvent]]:
+        collection, id_ = collection_and_id_from_fqid(event.fqid)
+
+        if collection != "mediafile":
+            return None
+
+        if isinstance(event, CreateEvent):
+            self.mediafiles[id_] = event.data
+        elif isinstance(event, UpdateEvent):
+            if id_ not in self.mediafiles:
+                self.mediafiles[id_] = self.new_accessor.get_model_ignore_deleted(
+                    event.fqid
+                )[0]
+            self.mediafiles[id_].update(event.data)
+        elif isinstance(event, ListUpdateEvent):
+            if id_ not in self.mediafiles:
+                self.mediafiles[id_] = self.new_accessor.get_model_ignore_deleted(
+                    event.fqid
+                )[0]
+            if remove_dict := event.remove:
+                for key, value in remove_dict.items():
+                    if not value:
+                        continue
+                    assert isinstance(
+                        old_value := self.mediafiles[id_].get(key), list
+                    ), f"'{event.fqid}' should have values for '{key}', because there is a ListUpdate.remove!"
```

Comment:
This is wrong, see https://github.com/OpenSlides/OpenSlides/wiki/Datastore-Service#interface: If a `remove` happens on a non-existent field, it is simply ignored.